File: 0001-Two_Sum/two_sum.cpp
#include <unordered_map>
#include <vector>
#include <iostream>

using namespace std;

std::vector<int> twoSum(std::vector<int>& nums, int target)
{
    std::unordered_map<int, int> map;

    for (int i = 0; i < nums.size(); i++) {
        if (map.count(nums[i]) > 0)
            return {map[nums[i]], i};

        map[target - nums[i]] = i;
    }

    return {};
}

int main()
{
    vector<int> arr = {2,7,11,15};
    int const& target = 9;
    arr = twoSum(arr, target);
    for (int const& it : arr) {
        cout << it << ' ';
        cout << endl;
    }
    return 0;
}

File: java/src/main/java/tjchiang/leetcode/L0001_Two_Sum/TwoSum.java
package tjchiang.leetcode.L0001_Two_Sum;

import java.util.HashMap;
import java.util.Map;

/**
 * <a href="https://leetcode.com/problems/two-sum/description/">1. Two Sum</a>
 */
public class TwoSum {

    public int[] solve(int[] nums, int target) {
        Map<Integer, Integer> map = new HashMap<>();
        for (int i = 0; i < nums.length; ++i) {
            int complement = target - nums[i];
            if (map.containsKey(complement)) {
                return new int[]{map.get(complement), i};
            }
            map.put(nums[i], i);
        }
        return new int[]{};
    }

}
